File: scripts/export_decompiled.py
# Ghidra Python script to export decompiled C code
# Run with: analyzeHeadless <project_dir> <project_name> -import <binary> -postScript export_decompiled.py

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

def run():
    program = currentProgram
    decompiler = DecompInterface()
    decompiler.openProgram(program)

    # Get output path from script arguments or use default
    output_dir = os.environ.get('GHIDRA_OUTPUT_DIR', '/tmp/ghidra_output')
    output_file = os.path.join(output_dir, program.getName() + '_decompiled.c')

    print("Decompiling {} to {}".format(program.getName(), output_file))

    with open(output_file, 'w') as f:
        f.write("// Decompiled from: {}\n".format(program.getExecutablePath()))
        f.write("// Using Ghidra {}\n\n".format(getGhidraVersion()))

        # Get all functions
        fm = program.getFunctionManager()
        functions = fm.getFunctions(True)

        func_count = 0
        for func in functions:
            if func.isExternal():
                continue

            results = decompiler.decompileFunction(func, 60, ConsoleTaskMonitor())

            if results and results.decompileCompleted():
                decomp_func = results.getDecompiledFunction()
                if decomp_func:
                    f.write("// Function: {} at {}\n".format(func.getName(), func.getEntryPoint()))
                    f.write(decomp_func.getC())
                    f.write("\n\n")
                    func_count += 1

        print("Decompiled {} functions".format(func_count))

    decompiler.dispose()

run()

File: scripts/export_decompiled.java
// Ghidra Java script to export decompiled C code
// @author asm2cpp
// @category Export

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.util.task.TaskMonitor;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

public class export_decompiled extends GhidraScript {

    @Override
    public void run() throws Exception {
        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(currentProgram);

        String outputDir = System.getenv("GHIDRA_OUTPUT_DIR");
        if (outputDir == null || outputDir.isEmpty()) {
            outputDir = "/tmp/ghidra_output";
        }

        File dir = new File(outputDir);
        if (!dir.exists()) {
            dir.mkdirs();
        }

        String outputFile = outputDir + "/" + currentProgram.getName() + "_decompiled.c";
        println("Decompiling " + currentProgram.getName() + " to " + outputFile);

        PrintWriter writer = new PrintWriter(new FileWriter(outputFile));
        writer.println("// Decompiled from: " + currentProgram.getExecutablePath());
        writer.println("// Using Ghidra\n");

        FunctionIterator functions = currentProgram.getFunctionManager().getFunctions(true);
        int funcCount = 0;

        while (functions.hasNext() && !monitor.isCancelled()) {
            Function func = functions.next();

            if (func.isExternal()) {
                continue;
            }

            DecompileResults results = decompiler.decompileFunction(func, 60, monitor);

            if (results != null && results.decompileCompleted()) {
                if (results.getDecompiledFunction() != null) {
                    writer.println("// Function: " + func.getName() + " at " + func.getEntryPoint());
                    writer.println(results.getDecompiledFunction().getC());
                    writer.println();
                    funcCount++;
                }
            }
        }

        writer.close();
        decompiler.dispose();
        println("Decompiled " + funcCount + " functions to " + outputFile);
    }
}
